File: docs/manual/scala/guide/cluster/code/docs/home/scaladsl/persistence/JdbcReadSideQuery.scala
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.scaladsl.persistence

//#imports
import scala.collection.immutable
import scala.collection.immutable.VectorBuilder
import akka.NotUsed
import com.lightbend.lagom.scaladsl.api.Service
import com.lightbend.lagom.scaladsl.api.ServiceCall
import com.lightbend.lagom.scaladsl.persistence.jdbc.JdbcSession

//#imports

trait JdbcReadSideQuery {

  trait BlogService extends Service {
    def getPostSummaries(): ServiceCall[NotUsed, immutable.IndexedSeq[PostSummary]]

    override def descriptor = ???
  }

  //#service-impl
  class BlogServiceImpl(jdbcSession: JdbcSession) extends BlogService {
    import JdbcSession.tryWith

    override def getPostSummaries() = ServiceCall { request =>
      jdbcSession.withConnection { connection =>
        tryWith(connection.prepareStatement("SELECT id, title FROM blogsummary")) { ps =>
          tryWith(ps.executeQuery()) { rs =>
            val summaries = new VectorBuilder[PostSummary]
            while (rs.next()) {
              summaries += PostSummary(rs.getString("id"), rs.getString("title"))
            }
            summaries.result()
          }
        }
      }
    }
    //#service-impl

  }
}

File: docs/manual/java/guide/cluster/code/docs/home/persistence/JdbcReadSideQuery.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.home.persistence;

// #imports
import akka.NotUsed;
import com.lightbend.lagom.javadsl.api.ServiceCall;
import com.lightbend.lagom.javadsl.persistence.jdbc.JdbcSession;
import org.pcollections.PSequence;
import org.pcollections.TreePVector;

import javax.inject.Inject;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
// #imports

public interface JdbcReadSideQuery {

  interface BlogService {
    public ServiceCall<NotUsed, PSequence<PostSummary>> getPostSummaries();
  }

  // #service-impl
  public class BlogServiceImpl implements BlogService {

    private final JdbcSession jdbcSession;

    @Inject
    public BlogServiceImpl(JdbcSession jdbcSession) {
      this.jdbcSession = jdbcSession;
    }

    @Override
    public ServiceCall<NotUsed, PSequence<PostSummary>> getPostSummaries() {
      return request -> {
        return jdbcSession.withConnection(
            connection -> {
              try (PreparedStatement ps =
                  connection.prepareStatement("SELECT id, title FROM blogsummary")) {
                try (ResultSet rs = ps.executeQuery()) {
                  PSequence<PostSummary> summaries = TreePVector.empty();

                  while (rs.next()) {
                    summaries =
                        summaries.plus(new PostSummary(rs.getString("id"), rs.getString("title")));
                  }

                  return summaries;
                }
              }
            });
      };
    }
  }
  // #service-impl

}
